sql/src/main/java/org/apache/druid/sql/calcite/external/ExternalTableMacro.java

```diff
@@ -59,6 +63,21 @@ public TranslatableTable apply(final List<Object> arguments)
       final InputFormat inputFormat = jsonMapper.readValue((String) arguments.get(1), InputFormat.class);
       final RowSignature signature = jsonMapper.readValue((String) arguments.get(2), RowSignature.class);
 
+      // Prevent a RowSignature that has a ColumnSignature with name "__time" and type that is not LONG because it will most
+      // likely  be used in a Calcite function, and cause an exception.
+      // Columns with name __time are automatically assumed to be of the type SqlTypeName.TIMESTAMP because of the logic
+      // present in RowSignatures#toRelDataType
+      // This is not the most optimal approach because of the following caveats:
+      // 1. __time in the external data source may map to a different column
+      // 2. __time in the external data source might be ignored while ingesting the data
+      // 3. Even if __time is of type LONG, it may be used with some function, which would cause the abovementioned Calcite failure
+      // The most optimal solution would be to prevent assumption of the __time column having the type SqlTypeName.TIMESTAMP
```

Comment:
Thanks for looking into this!

Is there a reason why we cannot do the more optimal solution now? Something like: for `__time`, have RowSignatures.toRelDataType return TIMESTAMP for regular Druid tables, and the type from the provided signature otherwise. To get the information through, we could add a parameter to toRelDataType like `boolean isDruidTable`.

Reply:
Let me take a go through this, but it should be possible. The thing preventing me from doing it was that I was unable to understand the reason for ignoring the RowSignature and casting the `__time` column to type TIMESTAMP.